kubernetes/conanfile.py

```diff
@@ -0,0 +1,70 @@
+from conan import ConanFile
+from conan.tools.cmake import CMakeToolchain, CMake, cmake_layout, CMakeDeps
+from conan.tools.files import copy
+from os.path import join
+
+class kubernetes_client_cRecipe(ConanFile):
+    name = "kubernetes_client_c"
+    version = "0.9.0"
+    package_type = "library"
+
+    # Optional metadata
+    license = "Apache-2.0"
+    url = "https://github.com/kubernetes-client/c"
+    description = "Official C client library for Kubernetes"
+    topics = ("kubernetes", "k8s", "kubernetes-client", "k8s-client")
+
+    # Binary configuration
+    settings = "os", "compiler", "build_type", "arch"
+    options = {"shared": [True, False], "fPIC": [True, False]}
+    default_options = {"shared": False, "fPIC": True}
+
+    # Sources are located in the same place as this recipe, copy them to the recipe
+    exports_sources = "config.h.in", "ConfigureChecks.cmake", "PreTarget.cmake", "PostTarget.cmake", "CMakeLists.txt", "src/*", "external/*", "api/*", "model/*", "include/*", "config/*", "watch/*", "websocket/*"
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            self.options.rm_safe("fPIC")
+
+    def configure(self):
+        if self.options.shared:
+            self.options.rm_safe("fPIC")
+
+    def layout(self):
+        cmake_layout(self)
+
+    def generate(self):
+        deps = CMakeDeps(self)
+        deps.generate()
+        tc = CMakeToolchain(self)
+        tc.generate()
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+    def package(self):
+        copy(self, "*.h", src=join(self.source_folder, "api"), dst=join(self.package_folder, "include/kubernetes/api"), keep_path=False)
+        copy(self, "*.h", src=join(self.source_folder, "model"), dst=join(self.package_folder, "include/kubernetes/model"), keep_path=False)
+        copy(self, "*.h", src=join(self.source_folder, "config"), dst=join(self.package_folder, "include/kubernetes/config"), keep_path=False)
+        copy(self, "*.h", src=join(self.source_folder, "include"), dst=join(self.package_folder, "include/kubernetes/include"), keep_path=False)
+        copy(self, "*.h", src=join(self.source_folder, "websocket"), dst=join(self.package_folder, "include/kubernetes/websocket"), keep_path=False)
+        copy(self, "*.h", src=join(self.source_folder, "external"), dst=join(self.package_folder, "include/kubernetes/external"), keep_path=False)
+        copy(self, "*.h", src=join(self.source_folder, "watch"), dst=join(self.package_folder, "include/kubernetes/watch"), keep_path=False)
+        cmake = CMake(self)
+        cmake.install()
+
+    def package_info(self):
+        self.cpp_info.libs = ["kubernetes"]
+
+    def requirements(self):
+        self.requires("libcurl/[^7.58]", headers=True, libs=True, transitive_headers=True)
+        self.requires("openssl/[~1.1]", headers=True, libs=True, transitive_headers=True)
```

Comment:
Can we specify a safe version of OpenSSL e.g. currently it is v3.3.0 as OpenSSL v1.1.1 reached its End of Life (EOL) on September 11, 2023.

Or we do not specify the OpenSSL version here and let users choose the version themselves.

Reply:
cannot specify openssl 3 because libwebsockets using openssl 1.1, which causes conflict.

I updated curl version range to be higher than 8 up to the latest.

Reply:
I think problem is in the libwebsockets package in conan, they explicitly stated for openssl requirement version 1.1.1w
https://github.com/conan-io/conan-center-index/blob/master/recipes/libwebsockets/all/conanfile.py
at line 228.

I will try to find a workaround in our conanfile.py to compile with openssl 3

Reply:
If you don't have a quick solution here, you can leave the version to the user's choice. For example `$openssl_version` in conanfile.py

Reply:
for our conanfile, solution is making openssl version forced `self.requires("openssl/[^3]", force=True)`, which also forces libwebsockets package to use openssl3 while building. 
But there is a compilation error when lws is built with openssl3, here is the full log of `conan create` command:
```shell
~/code/c/kubernetes (conan) $ conan create . --build=missing

======== Exporting recipe to the cache ========
kubernetes_client_c/0.9.0: Exporting package recipe: /home/yakup/code/c/kubernetes/conanfile.py
kubernetes_client_c/0.9.0: Copied 1 '.py' file: conanfile.py
kubernetes_client_c/0.9.0: Copied 3 '.cmake' files: PostTarget.cmake, ConfigureChecks.cmake, PreTarget.cmake
kubernetes_client_c/0.9.0: Copied 1 '.in' file: config.h.in
kubernetes_client_c/0.9.0: Copied 2 '.txt' files: CMakeLists.txt, CMakeLists.txt
kubernetes_client_c/0.9.0: Copied 644 '.h' files
kubernetes_client_c/0.9.0: Copied 644 '.c' files
kubernetes_client_c/0.9.0: Copied 1 '.licence' file: cJSON.licence
kubernetes_client_c/0.9.0: Exported to cache folder: /home/yakup/.conan2/p/kuber8a0c3df100384/e
kubernetes_client_c/0.9.0: Exported: kubernetes_client_c/0.9.0#35548bfa8c01cf3909331f9700a977fe (2024-04-25 08:27:45 UTC)

======== Input profiles ========
Profile host:
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=gnu17
compiler.libcxx=libstdc++11
compiler.version=11
os=Linux

Profile build:
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=gnu17
compiler.libcxx=libstdc++11
compiler.version=11
os=Linux


======== Computing dependency graph ========
Graph root
    cli
Requirements
    kubernetes_client_c/0.9.0#35548bfa8c01cf3909331f9700a977fe - Cache
    libcurl/8.6.0#357425be60d85553bfe65154194b71be - Cache
    libwebsockets/4.3.3#9b10666f1c07d726e9b193c16b0c5261 - Cache
    libyaml/0.2.5#67a8e4c752b3a452c4796ddc90589f74 - Cache
    openssl/3.2.1#c7b554068caae5eda12b735ea6f23d70 - Cache
    zlib/1.3.1#f52e03ae3d251dec704634230cd806a2 - Cache
Build requirements
    autoconf/2.71#f9307992909d7fb3df459340f1932809 - Cache
    automake/1.16.5#058bda3e21c36c9aa8425daf3c1faf50 - Cache
    gnu-config/cci.20210814#dc430d754f465e8c74463019672fb97b - Cache
    libtool/2.4.7#08316dad5c72c541ed21e039e4cf217b - Cache
    m4/1.4.19#b38ced39a01e31fef5435bc634461fd2 - Cache
    meson/1.2.2#04bdfb85d665c82b08a3510aee3ffd19 - Cache
    ninja/1.11.1#77587f8c8318662ac8e5a7867eb4be21 - Cache
    pkgconf/2.1.0#27f44583701117b571307cf5b5fe5605 - Cache
Resolved version ranges
    libcurl/[^8]: libcurl/8.6.0
    libwebsockets/[^4]: libwebsockets/4.3.3
    libyaml/[^0.2]: libyaml/0.2.5
    openssl/[^3]: openssl/3.2.1
    zlib/[>=1.2.11 <2]: zlib/1.3.1
Overrides
    openssl/[>=1.1 <4]: ['openssl/[^3]']
    openssl/1.1.1w: ['openssl/3.2.1#c7b554068caae5eda12b735ea6f23d70']

======== Computing necessary packages ========
kubernetes_client_c/0.9.0: Checking 11 compatible configurations
kubernetes_client_c/0.9.0: Compatible configurations not found in cache, checking servers
kubernetes_client_c/0.9.0: '823fa5212e9406cf95a84d608d1eebace0f37ab7': compiler.cppstd=98
kubernetes_client_c/0.9.0: '6720f5dfda5b3067abf0f97ab98f828f2f8fd150': compiler.cppstd=gnu98
kubernetes_client_c/0.9.0: 'd777cc226c6fa79eb8fdc7dbb3583b5ff4e960fb': compiler.cppstd=11
kubernetes_client_c/0.9.0: 'f92903f91300e3af82e9ac593568c1f179133bc3': compiler.cppstd=gnu11
kubernetes_client_c/0.9.0: 'ea094a27c81531fbe36137c25489e97acc378429': compiler.cppstd=14
kubernetes_client_c/0.9.0: '56888261f7b6a9060be5da0209674c2f8e45fd5a': compiler.cppstd=gnu14
kubernetes_client_c/0.9.0: '0e4f5e3e6341b116c5a448089fd25f7016fbbeaa': compiler.cppstd=17
kubernetes_client_c/0.9.0: '5ecc0e0569c3b5636f8ab1bb01df255052736d3e': compiler.cppstd=20
kubernetes_client_c/0.9.0: 'f75f9397019bcb13ac760f72cd37ee83be05f5ad': compiler.cppstd=gnu20
kubernetes_client_c/0.9.0: '6094b7cebe15ae9b67d260954f25cb49847b8227': compiler.cppstd=23
kubernetes_client_c/0.9.0: '7cb819605a3ee17a9e2b5e546effcab3b8f852a4': compiler.cppstd=gnu23
Requirements
    kubernetes_client_c/0.9.0#35548bfa8c01cf3909331f9700a977fe:c8ea2594cc602964a23453fa65ed75bb0dbfec4d - Build
    libcurl/8.6.0#357425be60d85553bfe65154194b71be:30324594e680f1c0c0a881247c2a0c20bd8733c0#dfb2eb0a4a82849a511c21a573ea0930 - Cache
    libwebsockets/4.3.3#9b10666f1c07d726e9b193c16b0c5261:f87c9aa87e38b55157a4f7a0556f301d2dd8ff9b - Build
    libyaml/0.2.5#67a8e4c752b3a452c4796ddc90589f74:b647c43bfefae3f830561ca202b6cfd935b56205#57519d3406af89fe6fce0abd0638fb7e - Cache
    openssl/3.2.1#c7b554068caae5eda12b735ea6f23d70:ba0c965640dfc9185194d683d05bfbe6419b47a3#d1726bfc4b66d9392fb7e73a90ed3d9c - Download (conancenter)
    zlib/1.3.1#f52e03ae3d251dec704634230cd806a2:b647c43bfefae3f830561ca202b6cfd935b56205#6b307bbcbae23635c4006543ffdbf3ef - Cache
Build requirements
Skipped binaries
    autoconf/2.71, automake/1.16.5, gnu-config/cci.20210814, libtool/2.4.7, m4/1.4.19, meson/1.2.2, ninja/1.11.1, pkgconf/2.1.0

======== Installing packages ========

-------- Downloading 1 package --------
openssl/3.2.1: Retrieving package ba0c965640dfc9185194d683d05bfbe6419b47a3 from remote 'conancenter' 
openssl/3.2.1: Package installed ba0c965640dfc9185194d683d05bfbe6419b47a3
openssl/3.2.1: Downloaded package revision d1726bfc4b66d9392fb7e73a90ed3d9c
libyaml/0.2.5: Already installed! (1 of 6)
zlib/1.3.1: Already installed! (2 of 6)

-------- Installing package libwebsockets/4.3.3 (4 of 6) --------
libwebsockets/4.3.3: Building from source
libwebsockets/4.3.3: Package libwebsockets/4.3.3:f87c9aa87e38b55157a4f7a0556f301d2dd8ff9b
libwebsockets/4.3.3: Copying sources to build folder
libwebsockets/4.3.3: Building your package in /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b
libwebsockets/4.3.3: Calling generate()
libwebsockets/4.3.3: Generators folder: /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/build/Release/generators
libwebsockets/4.3.3: Dependency library full path : /home/yakup/.conan2/p/opens8d497900e3fd5/p/lib/libssl.a
libwebsockets/4.3.3: Dependency library full path : /home/yakup/.conan2/p/opens8d497900e3fd5/p/lib/libcrypto.a
libwebsockets/4.3.3: CMakeToolchain generated: conan_toolchain.cmake
libwebsockets/4.3.3: CMakeToolchain generated: CMakePresets.json
libwebsockets/4.3.3: CMakeToolchain generated: ../../../src/CMakeUserPresets.json
libwebsockets/4.3.3: Generating aggregated env files
libwebsockets/4.3.3: Generated aggregated env files: ['conanbuild.sh', 'conanrun.sh']
libwebsockets/4.3.3: Calling build()
libwebsockets/4.3.3: Running CMake.configure()
libwebsockets/4.3.3: RUN: cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/build/Release/generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/home/yakup/.conan2/p/b/libwe8716b3ee4431c/p" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src"
-- Using Conan toolchain: /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/build/Release/generators/conan_toolchain.cmake
-- Conan toolchain: Setting CMAKE_POSITION_INDEPENDENT_CODE=ON (options.fPIC)
-- Conan toolchain: Setting BUILD_SHARED_LIBS = OFF
-- The C compiler identification is GNU 11.4.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Compiled with LWS_WITH_DIR and LWS_WITH_LEJP_CONF
Deselecting EVLIB_PLUGINS since building static
-- Found Git: /usr/bin/git (found version "2.34.1") 
fatal: not a git repository (or any of the parent directories): .git
Git commit hash: 
-- CMAKE_TOOLCHAIN_FILE='/home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/build/Release/generators/conan_toolchain.cmake'
-- Performing Test LWS_HAVE_inline
-- Performing Test LWS_HAVE_inline - Success
-- Performing Test LWS_HAVE___inline__
-- Performing Test LWS_HAVE___inline__ - Success
-- Performing Test LWS_HAVE___inline
-- Performing Test LWS_HAVE___inline - Success
-- Performing Test LWS_HAVE_MALLOC_TRIM
-- Performing Test LWS_HAVE_MALLOC_TRIM - Success
-- Performing Test LWS_HAVE_MALLOC_USABLE_SIZE
-- Performing Test LWS_HAVE_MALLOC_USABLE_SIZE - Success
-- Looking for fork
-- Looking for fork - found
-- Looking for getenv
-- Looking for getenv - found
-- Looking for malloc
-- Looking for malloc - found
-- Looking for memset
-- Looking for memset - found
-- Looking for realloc
-- Looking for realloc - found
-- Looking for socket
-- Looking for socket - found
-- Looking for strerror
-- Looking for strerror - found
-- Looking for vfork
-- Looking for vfork - found
-- Looking for execvpe
-- Looking for execvpe - found
-- Looking for getifaddrs
-- Looking for getifaddrs - found
-- Looking for snprintf
-- Looking for snprintf - found
-- Looking for _snprintf
-- Looking for _snprintf - not found
-- Looking for _vsnprintf
-- Looking for _vsnprintf - not found
-- Looking for getloadavg
-- Looking for getloadavg - found
-- Looking for atoll
-- Looking for atoll - found
-- Looking for _atoi64
-- Looking for _atoi64 - not found
-- Looking for _stat32i64
-- Looking for _stat32i64 - not found
-- Looking for clock_gettime
-- Looking for clock_gettime - found
-- Looking for localtime_r
-- Looking for localtime_r - found
-- Looking for gmtime_r
-- Looking for gmtime_r - found
-- Looking for ctime_r
-- Looking for ctime_r - found
-- Looking for getgrgid_r
-- Looking for getgrgid_r - found
-- Looking for getgrnam_r
-- Looking for getgrnam_r - found
-- Looking for getpwuid_r
-- Looking for getpwuid_r - found
-- Looking for getpwnam_r
-- Looking for getpwnam_r - found
-- Looking for timegm
-- Looking for timegm - found
-- Looking for in6addr.h
-- Looking for in6addr.h - not found
-- Looking for memory.h
-- Looking for memory.h - found
-- Looking for netinet/in.h
-- Looking for netinet/in.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stdlib.h
-- Looking for stdlib.h - found
-- Looking for strings.h
-- Looking for strings.h - found
-- Looking for string.h
-- Looking for string.h - found
-- Looking for sys/prctl.h
-- Looking for sys/prctl.h - found
-- Looking for sys/socket.h
-- Looking for sys/socket.h - found
-- Looking for sys/sockio.h
-- Looking for sys/sockio.h - not found
-- Looking for sys/stat.h
-- Looking for sys/stat.h - found
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for unistd.h
-- Looking for unistd.h - found
-- Looking for vfork.h
-- Looking for vfork.h - not found
-- Looking for sys/capability.h
-- Looking for sys/capability.h - not found
-- Looking for malloc.h
-- Looking for malloc.h - found
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for inttypes.h
-- Looking for inttypes.h - found
-- Looking for sys/resource.h
-- Looking for sys/resource.h - found
-- Looking for cap_set_flag in cap
-- Looking for cap_set_flag in cap - not found
-- Looking for 3 include files stdlib.h, ..., string.h
-- Looking for 3 include files stdlib.h, ..., string.h - found
-- Performing Test LWS_HAS_INTPTR_T
-- Performing Test LWS_HAS_INTPTR_T - Success
-- Performing Test LWS_HAS_PTHREAD_SETNAME_NP
-- Performing Test LWS_HAS_PTHREAD_SETNAME_NP - Success
-- Performing Test LWS_HAS_GETOPT_LONG
-- Performing Test LWS_HAS_GETOPT_LONG - Success
-- Performing Test LWS_HAVE_RTA_PREF
-- Performing Test LWS_HAVE_RTA_PREF - Success
-- Performing Test LWS_HAVE_SUSECONDS_T
-- Performing Test LWS_HAVE_SUSECONDS_T - Success
-- Performing Test LWS_HAVE_VISIBILITY
-- Performing Test LWS_HAVE_VISIBILITY - Success
-- Performing Test LWS_GCC_HAS_IGNORED_QUALIFIERS
-- Performing Test LWS_GCC_HAS_IGNORED_QUALIFIERS - Success
-- Performing Test LWS_GCC_HAS_TYPE_LIMITS
-- Performing Test LWS_GCC_HAS_TYPE_LIMITS - Success
-- Performing Test LWS_GCC_HAS_NO_DEPRECATED_DECLARATIONS
-- Performing Test LWS_GCC_HAS_NO_DEPRECATED_DECLARATIONS - Success
-- Looking for eventfd_read
-- Looking for eventfd_read - found
-- Performing Test LWS_HAVE_PIPE2
-- Performing Test LWS_HAVE_PIPE2 - Success
-- Performing Test LWS_HAVE_TCP_USER_TIMEOUT
-- Performing Test LWS_HAVE_TCP_USER_TIMEOUT - Success
Compiling with SSL support
OpenSSL include dir: /home/yakup/.conan2/p/opens8d497900e3fd5/p/include
OpenSSL libraries: /home/yakup/.conan2/p/opens8d497900e3fd5/p/lib/libssl.a;/home/yakup/.conan2/p/opens8d497900e3fd5/p/lib/libcrypto.a
-- Looking for openssl/ecdh.h
-- Looking for openssl/ecdh.h - not found
-- Looking for SSL_CTX_set1_param
-- Looking for SSL_CTX_set1_param - not found
-- Looking for SSL_set_info_callback
-- Looking for SSL_set_info_callback - not found
-- Looking for X509_VERIFY_PARAM_set1_host
-- Looking for X509_VERIFY_PARAM_set1_host - not found
-- Looking for X509_VERIFY_PARAM_set1_host
-- Looking for X509_VERIFY_PARAM_set1_host - not found
-- Looking for RSA_set0_key
-- Looking for RSA_set0_key - not found
-- Looking for X509_get_key_usage
-- Looking for X509_get_key_usage - not found
-- Looking for SSL_CTX_EVP_PKEY_new_raw_private_key
-- Looking for SSL_CTX_EVP_PKEY_new_raw_private_key - not found
-- Looking for SSL_CTX_get0_certificate
-- Looking for SSL_CTX_get0_certificate - not found
-- Looking for SSL_get0_alpn_selected
-- Looking for SSL_get0_alpn_selected - not found
-- Looking for SSL_set_alpn_protos
-- Looking for SSL_set_alpn_protos - not found
-- Looking for EVP_aes_128_cfb8
-- Looking for EVP_aes_128_cfb8 - not found
-- Looking for EVP_aes_128_cfb128
-- Looking for EVP_aes_128_cfb128 - not found
-- Looking for EVP_aes_192_cfb8
-- Looking for EVP_aes_192_cfb8 - not found
-- Looking for EVP_aes_192_cfb128
-- Looking for EVP_aes_192_cfb128 - not found
-- Looking for EVP_aes_256_cfb8
-- Looking for EVP_aes_256_cfb8 - not found
-- Looking for EVP_aes_256_cfb128
-- Looking for EVP_aes_256_cfb128 - not found
-- Looking for EVP_aes_128_xts
-- Looking for EVP_aes_128_xts - not found
-- Looking for EVP_aes_128_ofb
-- Looking for EVP_aes_128_ofb - not found
-- Looking for EVP_aes_128_ecb
-- Looking for EVP_aes_128_ecb - not found
-- Looking for EVP_aes_128_ctr
-- Looking for EVP_aes_128_ctr - not found
-- Looking for RSA_verify_pss_mgf1
-- Looking for RSA_verify_pss_mgf1 - not found
-- Looking for HMAC_CTX_new
-- Looking for HMAC_CTX_new - not found
-- Looking for EVP_PKEY_new_raw_private_key
-- Looking for EVP_PKEY_new_raw_private_key - not found
-- Looking for SSL_SESSION_set_time
-- Looking for SSL_SESSION_set_time - not found
-- Looking for EC_KEY_new_by_curve_name
-- Looking for EC_KEY_new_by_curve_name - not found
-- Performing Test LWS_HAVE_SSL_EXTRA_CHAIN_CERTS
-- Performing Test LWS_HAVE_SSL_EXTRA_CHAIN_CERTS - Failed
-- Performing Test LWS_HAVE_EVP_MD_CTX_free
-- Performing Test LWS_HAVE_EVP_MD_CTX_free - Failed
-- Performing Test LWS_HAVE_OPENSSL_STACK
-- Performing Test LWS_HAVE_OPENSSL_STACK - Success
-- Looking for ECDSA_SIG_set0
-- Looking for ECDSA_SIG_set0 - not found
-- Looking for BN_bn2binpad
-- Looking for BN_bn2binpad - not found
-- Looking for EVP_aes_128_wrap
-- Looking for EVP_aes_128_wrap - not found
-- Looking for EC_POINT_get_affine_coordinates
-- Looking for EC_POINT_get_affine_coordinates - not found
-- Looking for TLS_client_method
-- Looking for TLS_client_method - not found
-- Looking for TLSv1_2_client_method
-- Looking for TLSv1_2_client_method - not found
Searching for OpenSSL executable and dlls
OpenSSL executable: /usr/bin/openssl
 GENCERTS = 0
DIR /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/build/Release CMP /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/cmake
-- Configuring done
-- Generating done
-- Build files have been written to: /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/build/Release

libwebsockets/4.3.3: Running CMake.build()
libwebsockets/4.3.3: RUN: cmake --build "/home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/build/Release" -- -j16
[  1%] Creating build include dir
[  1%] Built target GENHDR
[  3%] Building C object lib/CMakeFiles/websockets.dir/plat/unix/unix-misc.c.o
[  3%] Building C object lib/CMakeFiles/websockets.dir/plat/unix/unix-caps.c.o
[  4%] Building C object lib/CMakeFiles/websockets.dir/plat/unix/unix-init.c.o
[  5%] Building C object lib/CMakeFiles/websockets.dir/plat/unix/unix-file.c.o
[  8%] Building C object lib/CMakeFiles/websockets.dir/plat/unix/unix-service.c.o
[  8%] Building C object lib/CMakeFiles/websockets.dir/plat/unix/unix-pipe.c.o
[  9%] Building C object lib/CMakeFiles/websockets.dir/plat/unix/unix-sockets.c.o
[ 11%] Building C object lib/CMakeFiles/websockets.dir/plat/unix/unix-fds.c.o
[ 11%] Building C object lib/CMakeFiles/websockets.dir/tls/tls.c.o
[ 12%] Building C object lib/CMakeFiles/websockets.dir/tls/tls-network.c.o
[ 15%] Building C object lib/CMakeFiles/websockets.dir/tls/openssl/openssl-tls.c.o
[ 15%] Building C object lib/CMakeFiles/websockets.dir/tls/tls-sessions.c.o
[ 16%] Building C object lib/CMakeFiles/websockets.dir/tls/openssl/openssl-x509.c.o
[ 17%] Building C object lib/CMakeFiles/websockets.dir/tls/openssl/openssl-ssl.c.o
[ 18%] Building C object lib/CMakeFiles/websockets.dir/tls/openssl/openssl-session.c.o
[ 20%] Building C object lib/CMakeFiles/websockets.dir/tls/tls-server.c.o
In file included from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets.h:664,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/core/./private-lib-core.h:144,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/plat/unix/unix-sockets.c:28:
/home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets/lws-genhash.h:85:18: error: field ‘ctx’ has incomplete type
   85 |         HMAC_CTX ctx;
      |                  ^~~
In file included from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets.h:664,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/core/./private-lib-core.h:144,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/plat/unix/unix-caps.c:28:
/home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets/lws-genhash.h:85:18: error: field ‘ctx’ has incomplete type
   85 |         HMAC_CTX ctx;
      |                  ^~~
In file included from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets.h:664,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/core/./private-lib-core.h:144,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/tls/tls.c:25:
/home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets/lws-genhash.h:85:18: error: field ‘ctx’ has incomplete type
   85 |         HMAC_CTX ctx;
      |                  ^~~
In file included from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets.h:664,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/core/./private-lib-core.h:144,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/tls/openssl/openssl-session.c:25:
/home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets/lws-genhash.h:85:18: error: field ‘ctx’ has incomplete type
   85 |         HMAC_CTX ctx;
      |                  ^~~
In file included from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets.h:664,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/core/./private-lib-core.h:144,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/plat/unix/unix-init.c:28:
/home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets/lws-genhash.h:85:18: error: field ‘ctx’ has incomplete type
   85 |         HMAC_CTX ctx;
      |                  ^~~
In file included from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets.h:664,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/core/./private-lib-core.h:144,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/tls/tls-server.c:25:
/home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets/lws-genhash.h:85:18: error: field ‘ctx’ has incomplete type
   85 |         HMAC_CTX ctx;
      |                  ^~~
In file included from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets.h:664,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/core/./private-lib-core.h:144,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/tls/openssl/openssl-x509.c:26:
/home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets/lws-genhash.h:85:18: error: field ‘ctx’ has incomplete type
   85 |         HMAC_CTX ctx;
      |                  ^~~
In file included from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets.h:664,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/core/./private-lib-core.h:144,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/plat/unix/unix-service.c:28:
/home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets/lws-genhash.h:85:18: error: field ‘ctx’ has incomplete type
   85 |         HMAC_CTX ctx;
      |                  ^~~
In file included from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets.h:664,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/core/./private-lib-core.h:144,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/plat/unix/unix-pipe.c:28:
/home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets/lws-genhash.h:85:18: error: field ‘ctx’ has incomplete type
   85 |         HMAC_CTX ctx;
      |                  ^~~
In file included from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets.h:664,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/core/./private-lib-core.h:144,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/plat/unix/unix-fds.c:28:
/home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets/lws-genhash.h:85:18: error: field ‘ctx’ has incomplete type
   85 |         HMAC_CTX ctx;
      |                  ^~~
In file included from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets.h:664,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/core/./private-lib-core.h:144,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/tls/tls-sessions.c:25:
/home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets/lws-genhash.h:85:18: error: field ‘ctx’ has incomplete type
   85 |         HMAC_CTX ctx;
      |                  ^~~
In file included from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets.h:664,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/core/./private-lib-core.h:144,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/tls/tls-network.c:25:
/home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets/lws-genhash.h:85:18: error: field ‘ctx’ has incomplete type
   85 |         HMAC_CTX ctx;
      |                  ^~~
In file included from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets.h:664,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/core/./private-lib-core.h:144,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/tls/openssl/openssl-tls.c:25:
/home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets/lws-genhash.h:85:18: error: field ‘ctx’ has incomplete type
   85 |         HMAC_CTX ctx;
      |                  ^~~
In file included from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets.h:664,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/core/./private-lib-core.h:144,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/plat/unix/unix-file.c:28:
/home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets/lws-genhash.h:85:18: error: field ‘ctx’ has incomplete type
   85 |         HMAC_CTX ctx;
      |                  ^~~
In file included from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets.h:664,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/core/./private-lib-core.h:144,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/plat/unix/unix-misc.c:28:
/home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets/lws-genhash.h:85:18: error: field ‘ctx’ has incomplete type
   85 |         HMAC_CTX ctx;
      |                  ^~~
In file included from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets.h:664,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/core/./private-lib-core.h:144,
                 from /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/lib/tls/openssl/openssl-ssl.c:25:
/home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/src/include/libwebsockets/lws-genhash.h:85:18: error: field ‘ctx’ has incomplete type
   85 |         HMAC_CTX ctx;
      |                  ^~~
gmake[2]: *** [lib/CMakeFiles/websockets.dir/build.make:216: lib/CMakeFiles/websockets.dir/tls/tls-sessions.c.o] Error 1
gmake[2]: *** Waiting for unfinished jobs....
gmake[2]: *** [lib/CMakeFiles/websockets.dir/build.make:272: lib/CMakeFiles/websockets.dir/tls/openssl/openssl-session.c.o] Error 1
gmake[2]: *** [lib/CMakeFiles/websockets.dir/build.make:174: lib/CMakeFiles/websockets.dir/plat/unix/unix-fds.c.o] Error 1
gmake[2]: *** [lib/CMakeFiles/websockets.dir/build.make:76: lib/CMakeFiles/websockets.dir/plat/unix/unix-caps.c.o] Error 1
gmake[2]: *** [lib/CMakeFiles/websockets.dir/build.make:286: lib/CMakeFiles/websockets.dir/tls/tls-server.c.o] Error 1
gmake[2]: *** [lib/CMakeFiles/websockets.dir/build.make:146: lib/CMakeFiles/websockets.dir/plat/unix/unix-service.c.o] Error 1
gmake[2]: *** [lib/CMakeFiles/websockets.dir/build.make:188: lib/CMakeFiles/websockets.dir/tls/tls.c.o] Error 1
gmake[2]: *** [lib/CMakeFiles/websockets.dir/build.make:132: lib/CMakeFiles/websockets.dir/plat/unix/unix-pipe.c.o] Error 1
gmake[2]: *** [lib/CMakeFiles/websockets.dir/build.make:244: lib/CMakeFiles/websockets.dir/tls/openssl/openssl-x509.c.o] Error 1
gmake[2]: *** [lib/CMakeFiles/websockets.dir/build.make:202: lib/CMakeFiles/websockets.dir/tls/tls-network.c.o] Error 1
gmake[2]: *** [lib/CMakeFiles/websockets.dir/build.make:230: lib/CMakeFiles/websockets.dir/tls/openssl/openssl-tls.c.o] Error 1
gmake[2]: *** [lib/CMakeFiles/websockets.dir/build.make:258: lib/CMakeFiles/websockets.dir/tls/openssl/openssl-ssl.c.o] Error 1
gmake[2]: *** [lib/CMakeFiles/websockets.dir/build.make:118: lib/CMakeFiles/websockets.dir/plat/unix/unix-file.c.o] Error 1
gmake[2]: *** [lib/CMakeFiles/websockets.dir/build.make:104: lib/CMakeFiles/websockets.dir/plat/unix/unix-init.c.o] Error 1
gmake[2]: *** [lib/CMakeFiles/websockets.dir/build.make:90: lib/CMakeFiles/websockets.dir/plat/unix/unix-misc.c.o] Error 1
gmake[2]: *** [lib/CMakeFiles/websockets.dir/build.make:160: lib/CMakeFiles/websockets.dir/plat/unix/unix-sockets.c.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:1258: lib/CMakeFiles/websockets.dir/all] Error 2
gmake: *** [Makefile:166: all] Error 2

libwebsockets/4.3.3: ERROR: 
Package 'f87c9aa87e38b55157a4f7a0556f301d2dd8ff9b' build failed
libwebsockets/4.3.3: WARN: Build folder /home/yakup/.conan2/p/b/libwe8716b3ee4431c/b/build/Release
ERROR: libwebsockets/4.3.3: Error in build() method, line 450
        cmake.build()
        ConanException: Error 2 while executing
```

it's because HMAC_CTX_new should exist in openssl3 but it cannot find it.
```
-- Looking for HMAC_CTX_new
-- Looking for HMAC_CTX_new - not found
```

This error only happens when openssl3 is built as static library. I'm still trying to figure out a solution to this.

For your suggestion, if user passes openssl 3.x version then libwebsockets dependency will still fail to build.

Reply:
I know. But we can submit this PR using this method first. Then when you figure out the final solution, you can create a new PR to address it.

Have you reported this issue in the libwebsockets community? I think they should keep library dependencies safe.

Reply:
Never give the value "1" for openssl_version in documents in this repo
```
"openssl_shared": [True, False]    # default: False
"openssl_version": ["1", "3"]    # default: "1"
"curl_version": ["7", "8"]    # default: "8"
···